File: python/122_best-time-to-buy-and-sell-stock-ii.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if len(prices) == 0:
            return 0

        profit = 0
        current_price = prices[0]

        for p in prices:
            if p > current_price:
                profit += p - current_price
            current_price = p

        return profit

File: go/122_best-time-to-buy-and-sell-stock-ii.go
// Peak Valley Approach
func maxProfit(prices []int) int {
	valley, peak := prices[0], prices[0]
	ans := 0

	for i := 0; i < len(prices)-1; i++ {
		for i < len(prices)-1 && prices[i] >= prices[i+1] {
			i++
		}
		valley = prices[i]
		for i < len(prices)-1 && prices[i] <= prices[i+1] {
			i++
		}
		peak = prices[i]
		ans += peak - valley
	}
	return ans
}

/*---------------------------------------------------------------------------------------*/

// One Pass
func maxProfit(prices []int) int {
	ans := 0
	for i := 0; i < len(prices)-1; i++ {
		if prices[i] < prices[i+1] {
			ans += prices[i+1] - prices[i]
		}
	}
	return ans
}